runtime/compiler/wasm/writer_test.go

```diff
@@ -0,0 +1,288 @@
+			Code: &Code{
+				Locals: []ValueType{
+					ValueTypeI32,
+				},
+				Instructions: []Instruction{
+					InstructionLocalGet{0},
+					InstructionLocalGet{1},
+					InstructionI32Add{},
+				},
+			},
+		},
+	}
+
+	err := w.writeFunctionSection(functions)
+	require.NoError(t, err)
+
+	require.Equal(t,
+		[]byte{
+			// section ID: Function = 3
+			0x3,
+			// section size: 2 (LEB128)
+			0x82, 0x80, 0x80, 0x80, 0x0,
+			// function count: 1
+			0x1,
+			// type ID of function: 0
+			0x0,
+		},
+		b.data,
+	)
+}
+
+func TestWasmWriter_writeCodeSection(t *testing.T) {
+
+	t.Parallel()
+
+	var b buf
+	w := WASMWriter{&b}
+
+	functions := []*Function{
+		{
+			// not used, just for testing
+			Name: "add",
+			// not used, just for testing
+			TypeID: 0,
+			Code: &Code{
+				Locals: []ValueType{
+					ValueTypeI32,
+				},
+				Instructions: []Instruction{
+					InstructionLocalGet{0},
+					InstructionLocalGet{1},
+					InstructionI32Add{},
+				},
+			},
+		},
+	}
+
+	err := w.writeCodeSection(functions)
+	require.NoError(t, err)
+
+	require.Equal(t,
+		[]byte{
+			// Section ID: Code = 10
+			0xa,
+			// section size: 15 (LEB128)
+			0x8f, 0x80, 0x80, 0x80, 0x0,
+			// function count: 1
+			0x1,
+			// code size: 9 (LEB128)
+			0x89, 0x80, 0x80, 0x80, 0x0,
+			// number of locals: 1
+			0x1,
+			// number of locals with this type: 1
+			0x1,
+			// local type: i32
+			0x7f,
+			// opcode: local.get, 0
+			0x20, 0x0,
+			// opcode: local.get 1
+			0x20, 0x1,
+			// opcode: i32.add
+			0x6a,
+			// opcode: end
+			0xb,
+		},
+		b.data,
+	)
+}
+
+func TestWasmWriter(t *testing.T) {
+
+	t.Parallel()
+
+	var b buf
+
+	w := WASMWriter{&b}
+
+	err := w.writeMagicAndVersion()
+	require.NoError(t, err)
+
+	err = w.writeTypeSection([]*FunctionType{
+		{
+			Params:  []ValueType{ValueTypeI32, ValueTypeI32},
+			Results: []ValueType{ValueTypeI32},
+		},
+	})
+	require.NoError(t, err)
+
+	functions := []*Function{
+		{
+			// not used, just for testing
+			Name:   "add",
+			TypeID: 0,
+			Code: &Code{
+				// not used, just for testing
+				Locals: []ValueType{
+					ValueTypeI32,
+				},
+				Instructions: []Instruction{
+					InstructionLocalGet{0},
+					InstructionLocalGet{1},
+					InstructionI32Add{},
+				},
+			},
+		},
+	}
+
+	err = w.writeFunctionSection(functions)
+	require.NoError(t, err)
+
+	err = w.writeCodeSection(functions)
+	require.NoError(t, err)
+
+	require.Equal(t,
+		[]byte{
+			0x0, 0x61, 0x73, 0x6d, 0x1, 0x0, 0x0, 0x0,
+			0x1, 0x87, 0x80, 0x80, 0x80, 0x0, 0x1, 0x60,
+			0x2, 0x7f, 0x7f, 0x1, 0x7f, 0x3, 0x82, 0x80,
+			0x80, 0x80, 0x0, 0x1, 0x0, 0xa, 0x8f, 0x80,
+			0x80, 0x80, 0x0, 0x1, 0x89, 0x80, 0x80, 0x80,
+			0x0, 0x1, 0x1, 0x7f, 0x20, 0x0, 0x20, 0x1,
+			0x6a, 0xb,
+		},
+		b.data,
+	)
+
+	require.Equal(t,
+		`(module
+  (type (;0;) (func (param i32 i32) (result i32)))
+  (func (;0;) (type 0) (param i32 i32) (result i32)
+    (local i32)
+    local.get 0
+    local.get 1
+    i32.add))
+`,
+		wasm2wat(b.data),
```

Comment:
👍  Nice to compare to the official library

Reply:
Note that the tool is not called to "compare": The wasm package produces a binary, and the `wasm2wat` tool is used to check if the written binary (WASM) is valid by parsing it and converting it to the text format (WAT).